From caebe78ac2328640f12774e56f44eae8fcef91e9 Mon Sep 17 00:00:00 2001
From: Jussi Kantell <jussi.kantell@intel.com>
Date: Wed, 28 Nov 2012 10:43:36 +0200
Subject: Dalvik: Align dlmalloc allocations to BOARD_MALLOC_ALIGNMENT

BZ: 62196

In JB dalvik compiles bionic/libc/upstream-dlmalloc/malloc.c
into libdvm.so directly and BOARD_MALLOC_ALIGNMENT is ignored.
This patch adds support for BOARD_MALLOC_ALIGNMENT into Dvm.mk.
Based on java.lang.System.arraycopy instrumentation
this fixes array mis-alignment issue which is one
of the reasons why Quadrant Memory score fluctuates.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I182dddcff019fc731aecbe95777f930096011a68
Orig-MCG-Change-Id: I30bd0d252b1512563f4f581f1011cbd0a170e9a6
Signed-off-by: Jussi Kantell <jussi.kantell@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/Dvm.mk |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/vm/Dvm.mk b/vm/Dvm.mk
index ee06308..3485b76 100644
--- a/vm/Dvm.mk
+++ b/vm/Dvm.mk
@@ -227,6 +227,11 @@ ifeq ($(WITH_COPYING_GC),true)
   LOCAL_SRC_FILES += \
 	alloc/Copying.cpp.arm
 else
+# To customize dlmalloc's alignment, set BOARD_MALLOC_ALIGNMENT in
+# the appropriate BoardConfig.mk file.
+  ifneq ($(BOARD_MALLOC_ALIGNMENT),)
+     LOCAL_CFLAGS += -DMALLOC_ALIGNMENT=$(BOARD_MALLOC_ALIGNMENT)
+  endif
   LOCAL_SRC_FILES += \
 	alloc/DlMalloc.cpp \
 	alloc/HeapSource.cpp \
-- 
1.7.4.1

